# Wrangling + Tf-Idf model

In [437]:
import pandas as pd

import warnings
#warnings.filterwarnings('ignore')

In [438]:
#data = pd.read_csv('data.csv', error_bad_lines=False, encoding="utf-8") # CSV old way
data = pd.read_json("data_5scheduler.json") # json new way

Look at the data:

In [439]:
print(data.describe(include="object"))

                title   identifier description  source instructors offered  \
count            4443         4443        4443    4443        4443    4443   
unique           3761         4234        3989       5        1231     139   
top     Senior Thesis  PHYS-178-KS              Pomona          []           
freq               56            2          95    1446        2023    1559   

       prerequisites corequisites  
count           4443         4443  
unique           716           28  
top                                
freq            3359         4411  


### Duplicates:
It looks like there is only 3989 unique course descripition so let's remove duplicates based on 'description' column.
There are also rows with empty descriptions, which are not helpful

In [440]:
print(len(data))
data = data.drop_duplicates(subset='description')
data = data[data["description"] != ""]
print(len(data))

4443
3988


In [441]:
data.head(10)

title   identifier  \
0                  Introduction to American Cultures  AMST-103-HM   
1                         Print and American Culture  AMST-115-HM   
2                               Hyphenated Americans  AMST-120-HM   
3    Life: Knowledge, Belief, and Cultural Practices  ANTH-110-HM   
4  Introduction to the Anthropology of Science an...  ANTH-111-HM   
5                                   War and Conflict  ANTH-115-HM   
6                                      Rationalities  ANTH-134-HM   
7                 A History of Landscape Photography  ARHI-131-HM   
8              Modern and Contemporary Art Practices   ART-002-HM   
9                                        Photography   ART-033-HM   

                                         description      source  credits  \
0  An interdisciplinary introduction to principal...  HarveyMudd      300   
1  Covers numerous developments in American print...  HarveyMudd      300   
2  A focus on the experience of immigrants in the...  HarveyMudd      300   
3  An exploration of cultural attitudes toward li...  HarveyMudd      300   
4  An introduction to science and technology as c...  HarveyMudd      300   
5  “The wings of the butterfly—that cause the hur...  HarveyMudd      300   
6  What does it mean to be rational? Does it mean...  HarveyMudd      300   
7  This course explores how photographic landscap...  HarveyMudd      300   
8  This class is an experimental lecture style ar...  HarveyMudd      300   
9  Approaching the medium from an artistic perspe...  HarveyMudd      300   

          instructors                  offered  \
0             [Staff]                            
1        [Anup Gampa]                            
2          [Balseiro]                            
3           [de Laet]                            
4  [Marianne De Laet]                            
5           [de Laet]                            
6           [de Laet]  Offered alternate years   
7           [Fandell]                            
8           [Fandell]                            
9           [Fandell]                            

                                       prerequisites corequisites  \
0                                                                   
1                                                                   
2                                                                   
3                                                                   
4                                                                   
5                                                                   
6  Any introductory course in anthropology or any...                
7                                                                   
8                                                                   
9                                         ART002 HM                 

   currently_offered  fee  
0              False    0  
1               True    0  
2              False    0  
3              False    0  
4               True    0  
5              False    0  
6              False    0  
7              False    0  
8              False    0  
9              False  150

### Tf-Idf with scikit-learn
[Description](https://monkeylearn.com/blog/what-is-tf-idf/)

[Usage](https://kavita-ganesan.com/tfidftransformer-tfidfvectorizer-usage-differences/#.Y1M42ezMJhF)

Here is an example of how Tf-Idf would work if our documents were the following 4 sentences:

In [442]:
import sklearn
from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np

corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',
]
vectorizer = TfidfVectorizer(use_idf=True)
vectors = vectorizer.fit_transform(corpus)
firstv = vectors[0]
df = pd.DataFrame(firstv.T.todense(), index=vectorizer.get_feature_names(), columns=["tfidf"])
df = df.sort_values(by=["tfidf"], ascending = False)
print("TfIdf values for the first sentence")
print(df)


TfIdf values for the first sentence
             tfidf
first     0.580286
document  0.469791
is        0.384085
the       0.384085
this      0.384085
and       0.000000
one       0.000000
second    0.000000
third     0.000000


In the example above we can see the importance of each word ranked for the first sentence `'this is the first document'`. So, for example the word `first` is important since it doesn't appear in any other document. The word `the` is not as important since it appears in all other documents. And the word `third` is not important at all since it doesn't even appear in the first document.

### Rank classes based on a given word
Function `tfidf_word(word, data)` takes in the word we are interested in and the data we are looking at. The function returns an updated dataframe with a new column `"score"` that gives each class a score of importance based on the input word. 

In [443]:
def tfidf_word(word, data_1):
    data_2 = data_1.copy() # since we don't want to be making changes to our original dataframe
    corpus = list(data_2.description)
    vectorizer = TfidfVectorizer(use_idf=True)
    vectors = vectorizer.fit_transform(corpus)

    score_for_word = []
    words = vectorizer.get_feature_names()
    try:
        index = words.index(word)
    except:
        print("'" + word + "'" + " is not mentioned in any course descriptions")
        return

    for i in range(0, len(corpus)):
        value = vectors[i].T.todense()[index]
        score_for_word.append(value)

    score_for_word = [float(i) for i in score_for_word] # type cast each score to a float

    data_2["score"] = score_for_word
    data_2 = data_2.sort_values(by=["score"], ascending = False)
    return data_2

For example, let's say we are interested in ranking all of the classes based on the word `computer`:

In [444]:
tfidf_word('computer', data).head(10)

title     identifier  \
132                       Computer Science Seminar    CSCI-181-HM   
525             Special Topics in Computer Science    CSCI-181-CM   
1490                   Computer Science Colloquium    CSCI-188-PO   
426     Introduction to Computational Neuroscience   BIOL-133L-KS   
1491               Computer Science Senior Seminar    CSCI-190-PO   
1060         Computational Physics and Engineering    PHYS-100-KS   
518               Fundamentals of Computer Science    CSCI-052-CM   
3358         Computational Physics and Engineering    PHYS-100-KS   
137                    Computer Science Colloquium    CSCI-195-HM   
106   Introduction to Biology and Computer Science  CSCI-005GR-HM   

                                            description            source  \
132   Advanced topics of current interest in compute...        HarveyMudd   
525   Selected topics in computer science. May be re...  ClaremontMckenna   
1490  Colloquium presentations and discussions of to...            Pomona   
426   This course provides computational skills for ...  ClaremontMckenna   
1491  Reading, discussion and presentation of resear...            Pomona   
1060  This course is a comprehensive introduction to...  ClaremontMckenna   
518   A solid foundation in functional programming, ...  ClaremontMckenna   
3358  This course is a comprehensive introduction to...           Scripps   
137   Oral presentations and discussions of selected...        HarveyMudd   
106   This course introduces fundamental concepts fr...        HarveyMudd   

      credits           instructors          offered  \
132         0               [Staff]  Fall and Spring   
525       100                    []     Occasionally   
1490        0     [Joseph C Osborn]   Each semester.   
426       100                    []       Every fall   
1491       25     [Joseph C Osborn]   Each semester.   
1060      100          [Scot Gould]     Every spring   
518       100                    []     Occasionally   
3358      100                    []     Every spring   
137        50  [Melissa E. O'Neill]  Fall and Spring   
106       300  [Wu, Bush (Biology)]             Fall   

                                          prerequisites corequisites  \
132                            Permission of instructor                
525                                                                    
1490  CSCI 051A PO , or CSCI 051G PO , or CSCI 051J ...                
426                                                                    
1491  Senior standing and two CSCI core courses (inc...                
1060                                                                   
518                                                                    
3358   PHYS033L KS  , PHYS034L KS ; or PHYS030L KS ,...                
137                            Juniors and seniors only                
106                                                                    

      currently_offered  fee     score  
132               False    0  0.442194  
525               False    0  0.431867  
1490               True    0  0.422261  
426               False    0  0.342191  
1491               True    0  0.334160  
1060               True    0  0.327428  
518               False    0  0.314859  
3358              False    0  0.311172  
137                True    0  0.276373  
106               False    0  0.257451

These are the first 10 instances of the classes that are most related to the word `computer` ranked in descending order (more related classes are on top). So, we could recomend a student who is interested in `computer`
 to take these classes.

Bellow are the outputs for fords `data, culture, activism, fiction, environment`

In [445]:
tfidf_word('data', data).head(10)

title   identifier  \
526                   Advanced Projects in Data Science    DS-180-CM   
1365  Data Analysis and Programming for the Life Sci...  BIOL-174-PO   
513                         Foundations of Data Science  CSCI-036-CM   
269                            Nonlinear Data Analytics  MATH-178-HM   
3254                       CS1: Intro to Python and Viz    MS-059-SC   
2974                                       Econometrics  ECON-125-SC   
580                           Accounting Data Analytics  ECON-160-CM   
110             Data Structures and Program Development  CSCI-070-HM   
523                         Introduction to Data Mining  CSCI-145-CM   
969                         Introduction to Data Mining  MATH-166-CM   

                                            description            source  \
526   This course allows teams of students to wrestl...  ClaremontMckenna   
1365  This course explores the analysis of big data ...            Pomona   
513   Data science is the interdisciplinary study of...  ClaremontMckenna   
269   Analysis of nonlinear large dynamic data inclu...        HarveyMudd   
3254  This is an introduction to computer programmin...           Scripps   
2974  Statistical techniques for testing economic mo...           Scripps   
580   This course will introduce students to the use...  ClaremontMckenna   
110   Abstract data types including priority queues ...        HarveyMudd   
523   Data mining is the process of discovering patt...  ClaremontMckenna   
969   Data mining is the process of discovering patt...  ClaremontMckenna   

      credits                          instructors                    offered  \
526       100                          [Jeho Park]                 Every year   
1365      100                   [Andre Cavalcanti]  Last offered spring 2019.   
513       100                       [Sarah Cannon]                 Every year   
269       300                                 [Gu]                       Fall   
3254        0                                   []                              
2974      100                     [Roberto Pedace]                              
580       100                       [George Batta]                 Every year   
110       300  [Melissa E. O'Neill, Erin Talvitie]            Fall and Spring   
523       100                  [Charles Griffiths]                 Every year   
969       100                                   []                 Every year   

                                          prerequisites corequisites  \
526                                                                    
1365  BIOL 040 PO and one of the following CSCI 005 ...                
513                                                                    
269   CSCI070 HM  and (CSCI140 HM  or MATH131 HM  or...                
3254                                                                   
2974                           ECON 101  and ECON 120 .                
580                                                                    
110   (CSCI060 HM  or CSCI042 HM ), and at least one...                
523                                                                    
969                                                                    

      currently_offered  fee     score  
526                True    0  0.491459  
1365               True    0  0.449229  
513                True    0  0.437465  
269               False    0  0.393787  
3254              False    0  0.389261  
2974               True    0  0.372383  
580                True    0  0.367099  
110                True    0  0.364790  
523                True    0  0.345841  
969               False    0  0.344338

In [446]:
tfidf_word('culture', data).head(10)

title    identifier  \
3153  Introduction to the Philosophy and History of ...   HIST-123-SC   
3183  Introduction to the Philosophy and History of ...   HMSC-123-SC   
903                               Film and Mass Culture    LIT-138-CM   
785      Culture and Society in Weimar and Nazi Germany  HIST-139E-CM   
4122                                    Popular Culture     MS-125-PZ   
3023  Literature and Popular Culture in the Antebell...  ENGL-143S-SC   
3726                      Visual Culture at the Margins   ASAM-171-PZ   
748                                    Cold War America   HIST-099-CM   
764                                    Cold War America   HIST-118-CM   
3187                 Critical Theory and Modern Culture   HMSC-136-SC   

                                            description            source  \
3153  This course will focus on some of the major wo...           Scripps   
3183  This course will focus on some of the major wo...           Scripps   
903   This course will examine film as art and as me...  ClaremontMckenna   
785   A study of the transformation of German cultur...  ClaremontMckenna   
4122  This course will cover a broad range of histor...            Pitzer   
3023  The years preceding the Civil War saw both the...           Scripps   
3726  This course will examine various forms of visu...            Pitzer   
748   The Cold War dramatically shaped the politics,...  ClaremontMckenna   
764   The Cold War dramatically shaped the politics,...  ClaremontMckenna   
3187  This course explores historical and contempora...           Scripps   

      credits     instructors           offered prerequisites corequisites  \
3153      100              []        Every year                              
3183      100              []                                                
903       100              []  Every third year                              
785       100              []  Every other year                              
4122        0              []                                                
3023      100              []  Every other year                              
3726        0              []                                                
748       100  [Lily Geismer]      Occasionally                              
764       100              []  Every other year                              
3187      100              []                                                

      currently_offered  fee     score  
3153              False    0  0.365540  
3183              False    0  0.362115  
903               False    0  0.297434  
785               False    0  0.258521  
4122              False    0  0.248049  
3023              False    0  0.245847  
3726              False    0  0.238080  
748                True    0  0.233963  
764               False    0  0.233417  
3187              False    0  0.230769

In [447]:
tfidf_word('activism', data).head(10)

title   identifier  \
1409  Latinx Social Movements: Identity, Power, and ...  CHST-136-CH   
3637  Political Activism, Social Movements and the P...  ANTH-138-PZ   
2748                          Art, Activism, Propaganda  ARHI-184-SC   
364                         Activism, Vocation, Justice  RLST-168-HM   
3710             Introduction to Asian American Studies  ASAM-101-PZ   
2536                     Social and Political Movements   SOC-075-PO   
3421                                  Remaking the Self  POLI-173-SC   
2540  Los Angeles Communities: Transformations, Ineq...   SOC-114-CH   
354                       Global Environmental Politics  POST-140-HM   
3169        African American Women in the United States  HIST-171-AF   

                                            description      source  credits  \
1409  Latin/o/a/xs have historically used grassroots...      Pomona      100   
3637  By examining contemporary issues, themes, and ...      Pitzer        0   
2748  Explores the intersection of art, political ac...     Scripps      100   
364   The histories of social change activism are fi...  HarveyMudd      150   
3710  Introduction to the field of Asian American St...      Pitzer      100   
2536  Can activism from below change society and pol...      Pomona      100   
3421  How do social movements change the world by ch...     Scripps      100   
2540  Use of case study approach to explore the inte...      Pomona      100   
354   Analyzes the political dynamics driving global...  HarveyMudd      300   
3169  This course explores the distinctive and diver...     Scripps      100   

                                            instructors  \
1409                                     [A. Zimmerman]   
3637                                                 []   
2748                                                 []   
364                                             [Dyson]   
3710                                   [Rosanna Simons]   
2536                                          [C. Beck]   
3421                                                 []   
2540  [Jeffrey D. Groves, Frank Lubbock Miller, Fran...   
354                                         [Steinberg]   
3169                                                 []   

                        offered  \
1409                 Fall 2020.   
3637                              
2748               Occasionally   
364                               
3710                              
2536    Last offered fall 2018.   
3421               Occasionally   
2540  Last offered spring 2018.   
354                               
3169                              

                                          prerequisites corequisites  \
1409                                                                   
3637                                                                   
2748                                                                   
364                               Instructor permission                
3710                                                                   
2536                                                                   
3421                                                                   
2540  Any course in Chicanx-Latinx Studies or Sociology                
354                                                                    
3169                                                                   

      currently_offered  fee     score  
1409              False    0  0.481129  
3637              False    0  0.389310  
2748              False    0  0.304916  
364               False    0  0.258028  
3710               True    0  0.223309  
2536              False    0  0.219529  
3421              False    0  0.198564  
2540               True    0  0.175725  
354               False    0  0.174303  
3169              False    0  0.172807

In [448]:
tfidf_word('fiction', data).head(10)

title    identifier  \
915                 Advanced Creative Writing    LIT-181-CM   
4424              Journalism in Latin America   SPAN-162-PZ   
838          Introduction to Creative Writing    LIT-031-CM   
3050          Introduction to Fiction Writing   ENGL-193-SC   
3051        Advanced Fiction Writing Workshop  ENGL-194S-SC   
214                          Fiction Workshop    LIT-035-HM   
607                 The Francophone Caribbean   FREN-115-CM   
2500  Post-Soviet Russian Culture and Society   RUSS-182-PO   
917                  Advanced Fiction Writing    LIT-183-CM   
889                European Modernist Fiction    LIT-122-CM   

                                            description            source  \
915   This is a class for the student who is serious...  ClaremontMckenna   
4424   Better than Fiction: Journalism in Latin America            Pitzer   
838   This course offers the chance to explore three...  ClaremontMckenna   
3050  This is an introductory course on writing shor...           Scripps   
3051  This advanced fiction workshop is intended for...           Scripps   
214   This course is designed as an introductory wor...        HarveyMudd   
607   A study of works of writers and artists from H...  ClaremontMckenna   
2500  The course explores the major changes in Russi...            Pomona   
917   This advanced fiction workshop is intended for...  ClaremontMckenna   
889   The first half of the 20th century produced an...  ClaremontMckenna   

      credits            instructors          offered prerequisites  \
915       100                     []       Every year                 
4424        0                     []                                  
838       100                     []     Occasionally                 
3050      100       [Leila Mansouri]                                  
3051        0                     []                                  
214       300  [Salvador Plascencia]  Fall and Spring                 
607       100                     []     Occasionally                 
2500      100    [Larissa V. Rudova]     Spring 2022.   RUSS 044 PO   
917       100       [Mary Gaitskill]       Every year                 
889       100                     []     Occasionally                 

     corequisites  currently_offered  fee     score  
915                            False    0  0.400743  
4424                           False    0  0.372122  
838                            False    0  0.365195  
3050                            True    0  0.287160  
3051                           False    0  0.277572  
214                             True    0  0.276736  
607                            False    0  0.274805  
2500                            True    0  0.258057  
917                             True    0  0.255142  
889                            False    0  0.239483

In [449]:
tfidf_word('environment', data).head(10)

title    identifier  \
1809  Food and the Environment in Asia and the Pacific  HIST-101F-PO   
2815                                      Microbiology  BIOL-168L-KS   
1047                              Environmental Ethics   PHIL-187-CM   
456                                       Microbiology  BIOL-168L-KS   
3904                   Critical Environmental Analysis     EA-150-PZ   
2559   Africa, the Environment, and the Global Economy   SOC-189H-PO   
3883                                     Urban Ecology     EA-098-PZ   
3616                     Global Environmental Conflict   ANTH-082-PZ   
3600           Native Americans and Their Environments   ANTH-012-PZ   
325                                  Topics in Physics   PHYS-080-HM   

                                            description            source  \
1809  A single question inspired this seminar: what ...            Pomona   
2815  In this fundamental microbiology course we wil...           Scripps   
1047  An exploration of human beings’ ethical relati...  ClaremontMckenna   
456   In this fundamental microbiology course we wil...  ClaremontMckenna   
3904  A seminar examination of how environmental iss...            Pitzer   
2559  Drawing on sociology and related disciplines, ...            Pomona   
3883  Urban ecology is a subfield of ecology that de...            Pitzer   
3616  This class uses the tools of anthropology and ...            Pitzer   
3600  This course will investigate the traditional i...            Pitzer   
325   An area of physics is studied, together with i...        HarveyMudd   

      credits         instructors                    offered  \
1809      100      [S. Yamashita]  Last offered spring 2018.   
2815      100                  []                 Every year   
1047      100                  []               Occasionally   
456       100  [Pete Chandrangsu]               Occasionally   
3904        0                  []                              
2559      100       [S. Stefanos]                 Each fall.   
3883      100  [Heather Campbell]                              
3616        0                  []                              
3600      100     [Sheryl Miller]                              
325       300   [Donnelly, Saeta]                              

                                          prerequisites corequisites  \
1809                                                                   
2815  BIOL043L KS  or BIOL040L KS ; BIOL044L KS ; CH...                
1047                                                                   
456                                                                    
3904                                                                   
2559                                                                   
3883                                                                   
3616                                                                   
3600                                                                   
325                                         PHYS051 HM                 

      currently_offered  fee     score  
1809              False    0  0.339513  
2815              False    0  0.287316  
1047              False    0  0.286515  
456                True    0  0.277818  
3904              False    0  0.257127  
2559              False    0  0.221758  
3883               True    0  0.206919  
3616              False    0  0.205580  
3600               True    0  0.199418  
325               False    0  0.196412

### Rank classes based on a given class
`tfidf_id(id, data)` will take in a identifier of some course (i.e. "PHIL-187-CM") and return courses that are most similar to the givern course. To do it, we will look at which words ranked highest for the given class and find other classes where same word also ranked highest.

Similarly to `tfidf_word`, we will add new columns that have the score of each class based on each word. In thins case we will have multiple new columns that have similarity scores for different word that were important in our input class.

In [450]:
def tfidf_word_helper(word, data1, prefix=""):
    data2 = data1.copy()
    corpus = list(data2.description)
    vectorizer = TfidfVectorizer(use_idf=True)
    vectors = vectorizer.fit_transform(corpus)

    score_for_word = []
    words = vectorizer.get_feature_names()
    try:
        index = words.index(word)
    except:
        print("'" + word + "'" + " is not mentioned in any course descriptions")
        return

    for i in range(0, len(corpus)):
        value = vectors[i].T.todense()[index]
        score_for_word.append(value)

    score_for_word = [float(i) for i in score_for_word] # type cast each score to a float

    data2["score"+prefix] = score_for_word
    return data2

In [451]:
def tfidf_id(id, data_):
    index_id = list(data_.index[data_["identifier"] == id]) # index of input class
    
    if len(index_id) == 0:
        print("Couldn't find a course " + id)
        return
    else:
        index_id = index_id[0]
    print(data_.loc[index_id, "identifier"])
    print(data_.loc[index_id, "description"])

    corpus = list(data_.description)
    vectorizer = TfidfVectorizer(use_idf=True)
    vectors = vectorizer.fit_transform(corpus)
    words = vectorizer.get_feature_names()

    scores_id = vectors[index_id].T.todense() # score values for our given class
    scores_id = [float(i) for i in scores_id] # type cast to float
    score_for_word = {}

    
    for i in range(0, len(words)):
        if scores_id[i] > 0.2: # We will count a word as relevant if it's score if more than 0.2 (arbitrary value, subject to change)
            score_for_word[words[i]] = [scores_id[i], i]
    score_for_word = {k: v for k, v in sorted(score_for_word.items(), key=lambda item: item[1][0], reverse=True)} #ordear dict by descending vals
    print("Most important words in course "+ id + " and their scores and indices")
    print(score_for_word)

    data__ = data_.copy() ## since we're making changes to our dataframe, we don't want to save these changes in the original dataframe
    new_cols = []
    for word in score_for_word:
        new_cols.append("score_" + word)
        data__ = tfidf_word_helper(word, data__, "_"+word) # tf idf word helper will help us add new columns with prefixes, same algo as regular tfdf_word()

    data__ = data__.sort_values(by=new_cols, ascending= False)
    return data__
    

In [452]:
# This one works
tfidf_id("AMST-120-HM", data).head(10)

AMST-120-HM
A focus on the experience of immigrants in the United States and Americans of diverse ethnic backgrounds, as reflected in literature and critical theory. The course will weave together works that treat the lives of immigrants and minority groups in the United States with examinations of such contemporary issues as bilingual education, the conditions of migrant workers, and children as cultural and linguistic interpreters for their parents. The intentionally broad and interdisciplinary nature of the course enables exploration of cultural identities, socio-economic status, and gender-specific roles.
Most important words in course AMST-120-HM and their scores and indices
{'immigrants': [0.26855708018539415, 7214], 'the': [0.20853700869471273, 14088]}


title   identifier  \
2364        The Politics of Immigration and Citizenship  POLI-046-PO   
2                                  Hyphenated Americans  AMST-120-HM   
3209  Italians as Guests and Hosts: Intercultural En...  ITAL-136-SC   
1192  Images of Immigration in Spanish Literature an...  SPAN-122-CM   
3927                      Criminalization of Immigrants    FS-016-PZ   
4260          US Immigration and Transnational Politics  POST-174-CH   
1789                              US Immigarion History  HIST-029-PO   
3748  Immigration from and#8220;The Tropicsand#8221;...  CHLT-120-PZ   
771         Asian American History, 1850 to the Present  HIST-125-CM   
4262                    Immigration and Race in America  POST-175-CH   

                                            description            source  \
2364  Examines immigration and citizenship politics ...            Pomona   
2     A focus on the experience of immigrants in the...        HarveyMudd   
3209  This course examines the phenomenon of exchang...           Scripps   
1192  From an interdisciplinary perspective, this co...  ClaremontMckenna   
3927  How did immigration and the U.S. - Mexico bord...            Pitzer   
4260  Examines the factors shaping the size and comp...            Pitzer   
1789  This introductory seminar examines the history...            Pomona   
3748  This class will focus on the immigration movem...            Pitzer   
771   This survey course examines the history of Asi...  ClaremontMckenna   
4262  America has long prided itself in being a nati...            Pitzer   

      credits            instructors                    offered  \
2364      100                [Staff]  Last offered spring 2019.   
2         300             [Balseiro]                              
3209      100                     []                              
1192      100                     []           Every other year   
3927      100  [Steffanie Guillermo]                              
4260        0                     []                              
1789      100                [Staff]                 Fall 2021.   
3748      100   [Michael J. Ballagh]                              
771       100                     []               Occasionally   
4262        0                     []                              

                 prerequisites corequisites  currently_offered  fee  \
2364                                                     False    0   
2                                                        False    0   
3209  ITAL 044  or equivalent.                           False    0   
1192                                                     False    0   
3927                                                      True    0   
4260                                                     False    0   
1789                                                     False    0   
3748                                                      True    0   
771                                                      False    0   
4262                                                     False    0   

      score_immigrants  score_the  
2364          0.324335   0.031481  
2             0.268557   0.208537  
3209          0.236742   0.114895  
1192          0.209550   0.122038  
3927          0.174580   0.220290  
4260          0.168014   0.195696  
1789          0.156974   0.243783  
3748          0.155281   0.150721  
771           0.153383   0.119104  
4262          0.152440   0.088778

In [453]:
# Doesn't work. Gives wrong word scores for some reason
tfidf_id("CSCI-036-CM", data).head(10)

CSCI-036-CM
Data science is the interdisciplinary study of the tools and theory behind using data to extract knowledge. It combines ideas from statistics, computer science, and particular domains in the hard and social sciences in order to make predictions and optimal decisions. This course covers the foundations of data science including the basics of how to structure, visualize, transform, and model data.
Most important words in course CSCI-036-CM and their scores and indices
{'climate': [0.5680111476749007, 2808], 'change': [0.4435747516841977, 2519], 'earth': [0.3197073895630546, 4660]}


title    identifier  \
65    Global Climate Change: Non-linearity, Irrevers...   CHEM-041-HM   
533                               Global Climate Change    EA-100L-KS   
532                               Global Climate Change     EA-100-KS   
2993                        Global Climate Change w/Lab    EA-100L-KS   
1728                                     Climate Change  GEOL-020G-PO   
544                             Modeling Climate Change   ECON-090-CM   
1367                              Global Change Biology  BIOL-189E-PO   
2434              Seminar: Psychology of Climate Change  PSYC-180C-PO   
324                                  Climate and Energy   PHYS-078-HM   
2433  Climate of Change: Climate Science, Psychology...  PSYC-180C-JT   

                                            description            source  \
65    Principles of the chemical and physical basis ...        HarveyMudd   
533   Introduction to the Earth Sciences with a focu...  ClaremontMckenna   
532   An introduction to the earth sciences, this co...  ClaremontMckenna   
2993  An introduction to the Earth Sciences, this co...           Scripps   
1728  An integrated perspective of Earth’s dynamic c...            Pomona   
544   A deep understanding of climate change require...  ClaremontMckenna   
1367  Global Climate Change Biology. Relying on scie...            Pomona   
2434  This seminar will explore psychological perspe...            Pomona   
324   Our climate’s dominant behavior is determined ...        HarveyMudd   
2433  This cross-disciplinary seminar will explore t...            Pomona   

      credits                                instructors  \
65        300  [Lisa M. Sullivan, Thomas David Donnelly]   
533       100                                         []   
532       100                                         []   
2993      100                                         []   
1728      100                            [W. McLaughlin]   
544       100                               [Mark Vinci]   
1367      100                 [F. Hanzawa, N. Karnovsky]   
2434      100                               [A. Pearson]   
324       300                                 [Donnelly]   
2433      100                               [A. Pearson]   

                        offered  \
65                         Fall   
533                  Every year   
532                  Every year   
2993                 Every year   
1728               Spring 2022.   
544            Every other year   
1367                 Fall 2021.   
2434  Last offered spring 2021.   
324                               
2433                 Fall 2021.   

                                          prerequisites corequisites  \
65                                                                     
533                                                                    
532                                                                    
2993  BIOL043L KS  and BIOL044L KS ; or BIOL040L KS ...                
1728                                                                   
544                                                                    
1367                                       BIOL 041E PO                
2434                                        PSYC 051 PO                
324                                                                    
2433                                                                   

      currently_offered  fee  score_climate  score_change  score_earth  
65                 True    0       0.654041      0.306454     0.000000  
533               False    0       0.579752      0.362195     0.261053  
532               False    0       0.568011      0.443575     0.319707  
2993              False    0       0.564488      0.352659     0.254180  
1728              False    0       0.529256      0.000000     0.446840  
544                True    0       0.453498      0.265611     0.127627  
1367              False    0       0.447997    

In [454]:
# Also doesn't work
tfidf_id("LIT-138-CM", data).head(10)

LIT-138-CM
This course will examine film as art and as medium in the context of the rise of 20th-century “mass culture.” We will take up such topics as the role of film in producing the ideas of “mass culture”, the cinematic representation of the “masses”, film as an instrument of the standardization of culture and as a mode of resistance to it, film and modernism, film and postmodernism, representations of fascism in cinema, and “subculture” considered as an effect of mass culture.
Most important words in course LIT-138-CM and their scores and indices
{'accounting': [0.46959270338428, 539], 'financial': [0.39176367047494914, 5729], 'the': [0.21810273324145757, 14088]}


title   identifier  \
3798                 Accounting and Finanicial Analysis  ECON-087-PZ   
575                        Financial Statement Analysis  ECON-154-CM   
577                                   Accounting Ethics  ECON-156-CM   
929               Financial Reporting and Communication   FIN-386-CM   
1546           Managerial Accounting Financial Analysis  ECON-117-PO   
935                        Advanced Accounting Analysis   FIN-440-CM   
572   Asset and Income Measurement (Intermediate Acc...  ECON-150-CM   
571   International Accounting, Taxation, and Transf...  ECON-149-CM   
579                      Accounting Theory and Research  ECON-159-CM   
923    Leadership Development in Finance and Accounting  FIN-301B-CM   

                                            description            source  \
3798  Examines the role of accounting information in...            Pitzer   
575   Combines finance and accounting in a user-orie...  ClaremontMckenna   
577   A case-method survey of ethical problems confr...  ClaremontMckenna   
929   This course will introduce students to the lan...  ClaremontMckenna   
1546  Examines the role of accounting information in...            Pomona   
935   The focus of this course is the connection bet...  ClaremontMckenna   
572   This course examines both conceptual foundatio...  ClaremontMckenna   
571   An introduction to global accounting, cross-li...  ClaremontMckenna   
579   An intensive study of the evolution and develo...  ClaremontMckenna   
923   Continuation of FIN301A CM - Leadership Develo...  ClaremontMckenna   

      credits                 instructors         offered prerequisites  \
3798        0                          []                                 
575       100           [Peter J McAniff]      Every year                 
577       100  [Gary Raymond Birkenbeuel]    Occasionally                 
929       100                          []      Every fall                 
1546      100         [Richard S. Savich]    Each spring.                 
935       100                          []    Every spring                 
572       100             [Gregory Lonzo]  Every semester                 
571       100        [Bernadette LeGrant]    Occasionally                 
579       100                          []      Every year                 
923        25                          []    Every Spring                 

     corequisites  currently_offered  fee  score_accounting  score_financial  \
3798                           False    0          0.566749         0.236409   
575                             True    0          0.505073         0.105341   
577                             True    0          0.492398         0.136930   
929                            False    0          0.469593         0.391764   
1546                            True    0          0.446066         0.186068   
935                            False    0          0.417751         0.000000   
572                             True    0          0.378170         0.157747   
571                             True    0          0.373339         0.000000   
579                            False    0          0.270099         0.000000   
923                            False    0          0.261592         0.000000   

      score_the  
3798   0.151862  
575    0.067668  
577    0.029320  
929    0.218103  
1546   0.119525  
935    0.074625  
572    0.101332  
571    0.100037  
579    0.192996  
923    0.000000